jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudwatch/logs/DownloadLogStream.kt

```diff
@@ -0,0 +1,151 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.cloudwatch.logs
+
+import com.intellij.icons.AllIcons
+import com.intellij.openapi.fileChooser.FileChooserFactory
+import com.intellij.openapi.fileChooser.FileSaverDescriptor
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.progress.Task
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.Messages
+import com.intellij.openapi.util.io.FileUtilRt
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withContext
+import software.amazon.awssdk.services.cloudwatchlogs.CloudWatchLogsClient
+import software.amazon.awssdk.services.cloudwatchlogs.model.GetLogEventsRequest
+import software.aws.toolkits.core.utils.error
+import software.aws.toolkits.core.utils.getLogger
+import software.aws.toolkits.jetbrains.services.cloudwatch.logs.actions.OpenStreamInEditor
+import software.aws.toolkits.jetbrains.services.cloudwatch.logs.actions.buildStringFromLogsOutput
+import software.aws.toolkits.jetbrains.utils.ApplicationThreadPoolScope
+import software.aws.toolkits.jetbrains.utils.getCoroutineUiContext
+import software.aws.toolkits.jetbrains.utils.notifyError
+import software.aws.toolkits.jetbrains.utils.notifyInfo
+import software.aws.toolkits.resources.message
+import java.io.File
+import java.nio.file.Files
+import java.time.Instant
+
+class LogStreamDownloadTask(project: Project, val client: CloudWatchLogsClient, val logGroup: String, val logStream: String) :
+    Task.Backgroundable(project, message("cloudwatch.logs.opening_in_editor", logStream), true),
+    CoroutineScope by ApplicationThreadPoolScope("OpenLogStreamInEditor") {
+    private val edt = getCoroutineUiContext()
+
+    override fun run(indicator: ProgressIndicator) = runBlocking {
+        // Default content load limit is 20MB, default per page is 1MB/10000 log entries. so we load MaxLength/1MB
+        // until we give up and prompt the user to save to file
+        val maxPages = FileUtilRt.getUserContentLoadLimit() / (1 * FileUtilRt.MEGABYTE)
+        val startTime = Instant.now()
+        val buffer = StringBuilder()
+        var index = 0
+        val request = GetLogEventsRequest
+            .builder()
+            .startFromHead(true)
+            .logGroupName(logGroup)
+            .logStreamName(logStream)
+            .endTime(startTime.toEpochMilli())
+        val getRequest = client.getLogEventsPaginator(request.build())
+        getRequest.stream().forEach {
+            indicator.checkCanceled()
+            buffer.append(it.events().buildStringFromLogsOutput())
+            index++
```

Comment:
nit: use `forEachIndexed`

Reply:
Apparently if we do `asSequence()` it works, changed